File: ch8/js/g-fork.js
import { MemorySaver } from "@langchain/langgraph";

// Assuming graph is already created and configured
const graph = new StateGraph().compile({ checkpointer: new MemorySaver() });

const config = { configurable: { thread_id: "1" } };

const history = await Array.fromAsync(graph.getStateHistory(config));
console.log("History states:", history.length);

// replay a past state
if (history.length >= 3) {
  const result = await graph.invoke(null, history[2].config);
  console.log("Replayed state result:", result);
}

File: ch8/py/g-fork.py
from langgraph.graph import StateGraph
from langgraph.checkpoint.memory import MemorySaver


def main():
    # Create a simple graph
    builder = StateGraph()
    # Add nodes and edges as needed
    graph = builder.compile(checkpointer=MemorySaver())

    config = {"configurable": {"thread_id": "1"}}

    history = [state for state in graph.get_state_history(config)]

    print("History states:", len(history))

    # replay a past state
    if len(history) >= 3:
        result = graph.invoke(None, history[2].config)
        print("Replayed state result:", result)


if __name__ == "__main__":
    main()
